### PR TITLE
Closes issue #5143 Adds support for while updating an Addon it requires new permissions.

### PR DESCRIPTION
**Pull request by Amejia481**
---

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- [ ] **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt [hunk 1]
```diff
@@ -100,6 +138,51 @@ class DefaultAddonUpdater(
...
+     */
+    override fun onUpdatePermissionRequest(
+        current: WebExtension,
+        updated: WebExtension,
+        newPermissions: List<String>,
+        onPermissionsGranted: (Boolean) -> Unit
+    ) {
+        val storage = StatusStorage()
+
+        val allowedPreviously = storage.getAddonStatus(applicationContext, addonId = updated.id)
```
Comment (psymoon) [inline components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt]: I need to look into this further but this seems prone to race condition

File: components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt [hunk 2]
```diff
@@ -237,6 +237,23 @@ class GeckoEngine(
...
+    /**
+     * See [Engine.onUpdatePermissionRequest].
+     */
+    override fun onUpdatePermissionRequest(
+        current: WebExtension,
+        updated: WebExtension,
+        newPermissions: List<String>,
+        onPermissionsGranted: (Boolean) -> Unit
+    ) {
+        webExtensionDelegate?.onUpdatePermissionRequest(
```
[the remark this answers is not in the document]
Comment (Amejia481) [reply components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt]: We could, but I'm not completely sure if we should as this is the same pattern that we are using everywhere where the  `delegate` is used like `installWebExtension` or `uninstallWebExtension`.

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtensionDelegate.kt [hunk 3]
```diff
@@ -104,12 +104,14 @@ interface WebExtensionDelegate {
...
      * @param current The current [WebExtension].
      * @param updated The update [WebExtension] that requires extra permissions.
+     * @param newPermissions Contains a list of all the new permissions.
      * @param onPermissionsGranted A callback to indicate if the new permissions from the [updated] extension
      * are granted or not.
      */
     fun onUpdatePermissionRequest(
         current: WebExtension,
         updated: WebExtension,
+        newPermissions: List<String>,
```
[the remark this answers is not in the document]
Comment (Amejia481) [reply components/concept/engine/src/main/java/mozilla/components/concept/engine/webextension/WebExtensionDelegate.kt]: Yes, we really don't need duplicates 👍

File: components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt [hunk 4]
```diff
@@ -127,7 +210,66 @@ class DefaultAddonUpdater(
...
+            putExtra(NOTIFICATION_EXTRA_ADDON_ID, ext.id)
+        }
+        val allowPendingIntent: PendingIntent =
+            PendingIntent.getBroadcast(applicationContext, 0, allowIntent, 0)
+
+        val title =
+            applicationContext.getString(R.string.mozac_feature_addons_updater_notification_title)
+        val contentText = applicationContext.getString(
+            R.string.mozac_feature_addons_updater_notification_content,
+            "Ublock"
```
Comment (Amejia481) [inline components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt]: Here I need a way to get an `add-on` from a `extension` to the the localize name.

File: components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (csadilek) [reply components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt]: Just adding my note here so we won't forget. As discussed, this method isn't needed. There's an inline comment on the 'PromptDelegate' which is where this should go...

[the remark this answers is not in the document]
Comment (Amejia481) [reply components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/GeckoEngine.kt]: 👍

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/Engine.kt [hunk 5]
```diff
@@ -160,6 +160,25 @@ interface Engine {
...
+     * new version. This requires user interaction as the updated extension will not be installed,
+     * until the user grants the new permissions.
+     *
+     * @param current The current [WebExtension].
+     * @param updated The update [WebExtension] that requires extra permissions.
+     * @param newPermissions Contains a list of all the new permissions.
+     * @param onPermissionsGranted A callback to indicate if the new permissions from the [updated] extension
+     * are granted or not.
+     */
+    fun onUpdatePermissionRequest(
```
Comment (csadilek) [inline components/concept/engine/src/main/java/mozilla/components/concept/engine/Engine.kt]: See comment above why we don't need to add this to the engine.

File: components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt [hunk 6]
```diff
@@ -36,6 +50,30 @@ interface AddonUpdater {
      */
     fun unregisterForFutureUpdates(addonId: String)
 
+    /**
+     * Try to perform an update on the given [addonId].
+     * @param addonId The unique id of the addon.
+     */
+    fun updateNow(addonId: String)
```
[the remark this answers is not in the document]
Comment (Amejia481) [reply components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt]: What do you think about just name it "update"?

File: components/concept/engine/src/main/java/mozilla/components/concept/engine/Engine.kt (hunk 5 above)
Comment (Amejia481) [reply components/concept/engine/src/main/java/mozilla/components/concept/engine/Engine.kt]: Fixed!

**Comment by Amejia481** [comment]
This is ready for review :)

File: samples/browser/src/main/java/org/mozilla/samples/browser/SampleApplication.kt [hunk 7]
```diff
@@ -39,6 +41,8 @@ class SampleApplication : Application() {
         components.engine.warmUp()
 
         try {
+            GlobalAddonManagerProvider.initialize(components.addonManager)
+            GlobalAddonUpdaterProvider.initialize(components.addonUpdater)
```
Comment (csadilek) [inline samples/browser/src/main/java/org/mozilla/samples/browser/SampleApplication.kt]: I know we need one of those for the worker, but don't think we need two? I think we should just pass the manager to the updater's constructor. The updater is the only one accessing the 'GlobalAddonManagerProvider' anyway. So, we can make that dependency explicit.

File: samples/browser/src/main/java/org/mozilla/samples/browser/SampleApplication.kt [hunk 8]
```diff
@@ -55,7 +59,8 @@ class SampleApplication : Application() {
                     _, sessionId ->
                         val selected = components.sessionManager.findSessionById(sessionId)
                         selected?.let { components.tabsUseCases.selectTab(it) }
-                }
+                },
+                onUpdatePermissionRequest = components.addonUpdater::onUpdatePermissionRequest
```
[the remark this answers is not in the document]
Comment (Amejia481) [reply samples/browser/src/main/java/org/mozilla/samples/browser/SampleApplication.kt]: 👍

File: components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt [hunk 9]
```diff
@@ -36,6 +52,30 @@ interface AddonUpdater {
...
+     * @param current The current [WebExtension].
+     * @param updated The update [WebExtension] that requires extra permissions.
+     * @param newPermissions Contains a list of all the new permissions.
+     * @param onPermissionsGranted A callback to indicate if the new permissions from the [updated] extension
+     * are granted or not.
+     */
+    fun onUpdatePermissionRequest(
+        current: WebExtension,
+        updated: WebExtension,
+        newPermissions: List<String>,
```
[the remark this answers is not in the document]
Comment (Amejia481) [reply components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt]: 👍

File: components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt [hunk 10]
```diff
@@ -100,6 +145,49 @@ class DefaultAddonUpdater(
...
+     * See [AddonUpdater.onUpdatePermissionRequest]
+     */
+    override fun onUpdatePermissionRequest(
+        current: WebExtension,
+        updated: WebExtension,
+        newPermissions: List<String>,
+        onPermissionsGranted: (Boolean) -> Unit
+    ) {
+        val allowedPreviously =
+            updateStatusStorage.isPreviouslyAllowed(applicationContext, updated.id)
```
[the remark this answers is not in the document]
Comment (Amejia481) [reply components/feature/addons/src/main/java/mozilla/components/feature/addons/update/AddonUpdater.kt]: Let's use `wasPreviouslyAllowed`

File: samples/browser/src/main/java/org/mozilla/samples/browser/SampleApplication.kt (hunk 7 above)
Comment (Amejia481) [reply samples/browser/src/main/java/org/mozilla/samples/browser/SampleApplication.kt]: Updated

**Comment by Amejia481** [comment]
bors r=csadilek,psymoon

**Pull request merged**